nft/src/main.sw

```diff
@@ -0,0 +1,474 @@
+    /// Returns the approved address
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized
+    // fn get_approved(token_id: u64) -> Option<Identity> {
+    //     require(storage.state != 0, InitError::NFTNotInitalized);
+
+    //     let meta_data: MetaData = storage.meta_data.get(token_id);
+    //     Option::Some(meta_data.approved)
+    // }
+
+    /// Returns the tokens owned by the address
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been intialized
+    fn get_tokens(identity: Identity) -> u64 {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+        storage.owners.get(identity)
+    }
+
+    /// Returns the total supply for the NFT contract
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized
+    fn get_total_supply() -> u64 {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+        storage.token_supply
+    }
+
+    /// Returns whether the address is approved for all tokens
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized
+    fn is_approved_for_all(owner: Identity, operator: Identity) -> bool {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+
+        let hash: b256 = get_owner_operator_hash(owner, operator);
+        storage.operator_approval.get(hash)
+    }
+
+    /// Mints an NFT
+    ///
+    /// # Panics
+    /// 
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized 
+    /// - The amount is set to 0
+    /// - More NFTs than supply is minted
+    /// - The to Identity is not valid
+    /// - The sender is not approved to mint
+    /// - The sender sent the wrong asset
+    /// - The sender did not pay enough tokens
+    fn mint(to: Identity, amount: u64) -> bool {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+        require(amount != 0, InputError::MintAmountCannotBeZero);
+        require(
+            storage.token_supply >= (storage.token_count + amount), 
+            InputError::NotEnoughTokensToMint
+        );
+        require_identity_is_valid(to);
+
+        let sender: Identity = unwrap_identity(msg_sender());
+
+        /// Ensure that the sender is on the approved mint list
+        require(
+            !storage.access_control || 
+            (storage.access_control && storage.allowed_minters.get(sender)), 
+            AccessError::SenderDoesNotHaveAccessControl
+        );
+
+        /// Mint as many tokens as the sender has paid for
+        let mut i = 0;
+        while i < amount {
+            /// Create a new token id in sequential order
+            let token_id: u64 = storage.token_count + 1;
+            let empty_identity: Identity = Identity::Address(~Address::from(NATIVE_ASSET_ID));
```

Comment:
```suggestion
            let empty_identity = ~Address::from(NATIVE_ASSET_ID);
```

Shouldn't this work?

Reply:
Also, consider where this is used. If the 0x000 address becomes a contract that can perform work then this may be an issue. There has been some mention of this

Reply:
Not sure what I would use in place of the 0 address. As far as I'm aware Sway or Rust doesn't have a `null` in the way that C++ does. In solidity the 0 address is reserved. If a contract were to receive the 0 address that would let them do a lot of damage.